docs/design/security.md

```diff
@@ -0,0 +1,150 @@
+# GKM Security Model
+
+The GPU Kernel Manager (GKM) enforces a multi-layered security model to ensure
+that only trusted, signed GPU kernel cache images are used, that workloads
+consume validated resources, and that system components maintain strong
+ownership boundaries.
+
+## Image Signature Verification and Digest Trust
+
+To guarantee the integrity and provenance of kernel cache images, GKM uses
+a cosign-based signature verification pipeline enforced via Kubernetes
+admission and controller logic.
+
+> **Note:** Integration with projects like Kyverno will be considered in the
+> future when it supports image digest mutation for Custom Resource Objects.
+
+### Components and Roles
+
+<!-- markdownlint-disable  MD013 -->
+<!-- Temporarily disable MD013 - Line length to keep the table formatting  -->
+| Component                  | Responsibilities                                                                 |
+|----------------------------|-----------------------------------------------------------------------------------|
+| **Admission Webhook (Combined)** | - Verifies signatures<br>- Resolves image tag to digest<br>- Mutates trusted annotation<br>- Validates annotation tampering |
+| **GKM Operator**           | - Promotes trusted digest to `.status.resolvedDigest`<br>- Sets `.status.lastUpdated` |
+| **GKM Agent**              | - Watches `.status.resolvedDigest`<br>- Pulls image by digest<br>- Validates compatibility<br>- Updates `GKMCacheNode` status |
+
+<!-- markdownlint-enable  MD013 -->
+
+### Workflow
+
+#### User Action
+
+User creates or updates a GKMCache or ClusterGKMCache with a new spec.image
+(e.g., `quay.io/cache:latest`)
+
+#### Combined Validation Admission Webhook Executes
+
+**Mutation Phase (on CREATE or UPDATE)**:
+
+- Resolves the tag to a content digest (`sha256:...`).
+- Verifies the image signature using
+  [cosign](https://github.com/sigstore/cosign).
+- Adds a protected annotation to the CR:
+
+```yaml
+metadata:
+  annotations:
+    gkm.io/resolvedDigest: sha256:abc123...
+```
+
+> **Note:** This annotation is considered *trusted system state*.
+> The **Validating Admission Webhook** must be configured to
+> reject any attempt by a user to add, modify, or remove this annotation.
+> Only the GKM webhook or controller is authorized to manage this field.
+
+**Validation Phase**:
+
+Denies the request if:
+
+- The image is not signed or verification fails
+- The user attempts to:
+  - Modify the `gkm.io/resolvedDigest` annotation directly
+- Only trusted service accounts (e.g. the webhook itself) can mutate
+  this annotation.
+
+### Operator Flow
+
+The **GKM Operator**:
+
+- Watches `GKMCache` and `ClusterGKMCache` CRs for new or changed `spec.image`
+  or annotations.
+- When it finds a trusted annotation, it updates the CR `status`:
```

Comment:
Why the two step process: WebHook -> Operator? Can WebHook not update `status.resolvedDigest`? Or can Agent not just read `gkm.io/resolvedDigest` annotation (which is preferred so that there are not two writers of `status`)?

Reply:
> Why the two step process: WebHook -> Operator? Can WebHook not update status.resolvedDigest? Or can Agent not just read gkm.io/resolvedDigest annotation (which is preferred so that there are not two writers of status)?

The webhook can only update spec fields. One option it to move the `resolvedDigest` to spec and then just make sure that a user can't modify that field. But I thought we both thought that `resolvedDigest` belonged more in Status.
The Operator will need to update the modification time anyway - so I thought 2 birds one stone?